### PR TITLE
Gateway: Handling of 'Sharding Required', Proper Closures and General Cleanup

### PR DESCRIPTION
**Pull request by mavolin**
### Sharding
This PR adds support for handling the `4011` aka "Sharding Required" close code, to allow using `Gateway` for sharding. Handling is provided through a `OnShardingRequired` function inside the `Gateway` struct, called if Discord closes a connection with a `4011` error code. Additionally, the Gateway now won't continue to attempt to reconnect, but instead exit.

### `Gateway.Close`
`Gateway.Close` was adapted to behave like `.CloseGracefully` in an effort to ensure, that a user made call to `.Close` will end the WS connection, even if `Gateway` is in the process of reconnecting. Consequently, `.CloseGracefully` has been deprecated, and it has been advised to use `.Close` instead. Furthermore, `.Reconnect` and `.ReconnectCtx` were unexported, as user-triggered reconnections are now impossible.

### Other Improvements
* When attempting to reconnect, `Gateway.reconnectCtx` will now wait for an increasing delay between each attempt to connect, with a max delay of 60 seconds.
* `Gateway.ReconnectTimeout` was deprecated in favor of the new `Gateway.ReconnectAttempts`, to have more control over retries despite delays between tries

### CODE REVIEW

**Review by diamondburned — APPROVED** [review]

File: gateway/gateway.go [hunk 1]
```diff
@@ -116,12 +128,18 @@ type Gateway struct {
 	// Defaults to noop.
 	FatalErrorCallback func(err error)
 
+	// OnScalingRequired is the function called, if discord closes with error
+	// code 4011 aka Scaling Required.
+	OnScalingRequired func()
```
Comment (diamondburned) [inline gateway/gateway.go:137]: Is this only called on initial identification or on reconnects as well?

File: gateway/gateway.go [hunk 2]
```diff
@@ -246,56 +283,62 @@ func (g *Gateway) SessionID() string {
 	return g.sessionID
 }
 
-// Reconnect tries to reconnect until the ReconnectTimeout is reached, or if
+// reconnect tries to reconnect until the ReconnectTimeout is reached, or if
```
Comment (diamondburned) [inline gateway/gateway.go]: This is breaking the API; why the change?

Comment (diamondburned) [reply gateway/gateway.go]: There are also changes on the contextful version of this method.

File: utils/wsutil/ws.go [hunk 3]
```diff
@@ -157,50 +157,37 @@ func (ws *Websocket) SendCtx(ctx context.Context, b []byte) error {
...
-	return err
+
+	if graceful {
+		if gc, ok := ws.conn.(GracefulCloser); ok {
+			WSDebug("Conn: Closing gracefully")
+			return gc.CloseGracefully()
+		}
+
+		WSDebug("Conn: The Websocket's Connection does not support graceful closure. Closing normally instead.")
+		return ws.conn.Close()
```
Comment (diamondburned) [inline utils/wsutil/ws.go]: This like can just not be here; execution will fall out into the `Close()` after it. The redundant debug logging is fine.

**Review by diamondburned — CHANGES_REQUESTED** [review]

File: gateway/gateway.go (hunk 2 above)
Comment (mavolin) [reply gateway/gateway.go]: Because `Reconnect` does not work anymore, since Closures, as discussed, invalidate the session ID.

File: gateway/gateway.go (hunk 1 above)
Comment (mavolin) [reply gateway/gateway.go:137]: This is called whenever Discord closes the WS with 4011, not afterwards. On call, the Gateway will be closed, and the user should signal all other shards to disconnect, and reconnect with more shards.

File: gateway/gateway.go (hunk 2 above)
Comment (diamondburned) [reply gateway/gateway.go]: Closes aren't specifically relevant in the context of the caller; the function should automatically handle closing (ungracefully) and resume as normally.

---

For the record, I still think there should be a way to close without invalidating the session ID. I think a new method, maybe `Terminate`, should be added. The old `CloseGracefully` can call `Close` but marked as deprecated.

I'm not for breaking the API without bumping the major version.

File: utils/wsutil/ws.go (hunk 3 above)
Comment (mavolin) [reply utils/wsutil/ws.go]: I don't understand, please elaborate.

File: gateway/gateway.go (hunk 2 above)
Comment (mavolin) [reply gateway/gateway.go]: Alright, will implemt this tomorrow.

Comment (mavolin) [reply gateway/gateway.go]: However, I think `Pause` would be more self-explanatory, as it emphasizes that the connection can be resumed, which not be so apparent when using `Terminate`. What do you think?

File: utils/wsutil/ws.go (hunk 3 above)
Comment (diamondburned) [reply utils/wsutil/ws.go]: Like this:

```go
	if graceful {
		if gc, ok := ws.conn.(GracefulCloser); ok {
			WSDebug("Conn: closing gracefully")
			return gc.CloseGracefully()
		}

		WSDebug("Conn: graceful closure unsupported; closing normally instead.")
	}

	WSDebug("Conn: closing")
	return ws.conn.Close()
```

Comment (mavolin) [reply utils/wsutil/ws.go]: Sure, I just wanted to prevent double logging, but I guess `WSDebug` is verbose enough as for this to not matter. Again, I will do this tomorrow

File: gateway/gateway.go (hunk 2 above)
Comment (diamondburned) [reply gateway/gateway.go]: Sounds good to me.

**Review by diamondburned — APPROVED** [review]
